chainstate/src/detail/tokens/mod.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common::{
+    chain::{
+        config::{TOKEN_MAX_DEC_COUNT, TOKEN_MAX_TICKER_LEN, TOKEN_MAX_URI_LEN},
+        tokens::{TokenData, TokensError},
+        Block, Transaction,
+    },
+    primitives::{Amount, Id, Idable},
+};
+use utils::ensure;
+
+pub fn check_tokens_transfer_data(
+    source_block_id: Id<Block>,
+    tx: &Transaction,
+    amount: &Amount,
+) -> Result<(), TokensError> {
+    // Check amount
+    ensure!(
+        amount > &Amount::from_atoms(0),
+        TokensError::TransferZeroTokens(tx.get_id(), source_block_id)
+    );
+
+    Ok(())
+}
+
+pub fn check_tokens_burn_data(
+    tx: &Transaction,
+    source_block_id: &Id<Block>,
+    amount_to_burn: &Amount,
+) -> Result<(), TokensError> {
+    // Check amount
+    ensure!(
+        amount_to_burn != &Amount::from_atoms(0),
+        TokensError::BurnZeroTokens(tx.get_id(), *source_block_id)
+    );
+    Ok(())
+}
+
+pub fn check_tokens_issuance_data(
+    token_ticker: &Vec<u8>,
+    amount_to_issue: &Amount,
+    number_of_decimals: &u8,
+    metadata_uri: &Vec<u8>,
+    tx_id: Id<Transaction>,
+    source_block_id: Id<Block>,
+) -> Result<(), TokensError> {
+    // Check token name
+    if token_ticker.len() > TOKEN_MAX_TICKER_LEN
+        || token_ticker.is_empty()
+        || !String::from_utf8_lossy(token_ticker).is_ascii()
```

Comment:
Never, ever, use lossy conversions in consensus. This is a recipe for disaster. Basically people can post things in the token ticker that are not detectable by our software.

The right way to do this check if with `from_utf8()`, then do error handling.

Also double-check that the result has only alphanumeric chars with `result.chars().all(char::is_alphanumeric)`.

Reply:
Also add tests that ensure that token tickers can't have symbols after implementing this.

Reply:
I've got you. I fixed it. Also, `token_issue_test` has the check that tickers can't have symbols.

Reply:
I will push all updates later, because my credential expired on github